pandas/core/indexes/base.py

```diff
@@ -7156,6 +7155,43 @@ def shape(self) -> Shape:
         return (len(self),)
 
 
+def maybe_sequence_to_range(sequence) -> Any | range:
+    """
+    Convert a 1D sequence to a range if possible.
+
+    Returns the input if not possible.
+
+    Parameters
+    ----------
+    sequence : 1D sequence
+    names : sequence of str
+
+    Returns
+    -------
+    Any : input or range
+    """
+    if hasattr(sequence, "dtype") and not isinstance(sequence, np.ndarray):
```

Comment:
I find the duck typing here somewhat non-obvious - maybe comments will help? Is this first branch supposed to be for extension types?

Reply:
> Is this first branch supposed to be for extension types? 

Essentially yes, I think I can use a more specific check here

Reply:
Updated the check here to be more explicit (exclude Series and Index objects)